fastpy/fastpy/run/Runner.py

```diff
@@ -0,0 +1,99 @@
+import os
+import shutil
+import itertools
+from subprocess import call
+
+from fastpy.io.config import load_json_config, store_json_config
+from fastpy.utils import get_date_time_tag
+from common import DATA_DIR_PATH, PROJECT_ROOT_PATH
+
+# This mapping will be used to pass command line arguments to benchmark executable
+PARAM_TO_C_MAP = {'algorithm':  '-a',
+                  'obj_func':   '-o',
+                  'dimension':  '-d',
+                  'n_iter':     '-n',
+                  'population': '-p',
+                  'min_val':    '-y',
+                  'max_val':    '-z'}
+
+BENCHMARK_BIN_DIR = os.path.join(PROJECT_ROOT_PATH, '../benchmark')
+BENCHMARK_BIN = 'benchmark'
+
+TIMING_OUT_FILE = 'timings.csv'
+SOLUTION_OUT_FILE = 'solution.csv'
+
+
+class BenchmarkRunner:
+
+    def __init__(self, config_path, data_dir=DATA_DIR_PATH):
+        self.config = load_json_config(config_path)
+        self.data_dir = data_dir
+        self.timestamp = get_date_time_tag()
+
+    def run_benchmarks(self):
+        """Main wrapper function to loop over all possible parameter combinations."""
+
+        os.mkdir(self._output_dir)  # main output dir for this whole run
+        shutil.copy(os.path.join(BENCHMARK_BIN_DIR, BENCHMARK_BIN), self._output_dir)
+        store_json_config(self.config, self._output_dir, 'config.json')
+
+        for run_idx, run_config in enumerate(self._build_param_sets()):
+
+            sub_dir = os.path.join(self._output_dir, f'run_{run_idx}')  # sub output dir for one param combo
+            os.mkdir(sub_dir)
+
+            store_json_config(run_config, sub_dir, 'run_config.json')
+
+            self._run_algorithm(run_config, sub_dir)
+
+    def _run_algorithm(self, run_config, sub_dir):
+        """Subprocess call to run a single algorithm."""
+
+        call_str = ' '.join([self._benchmark_bin, self._create_params_str(run_config)])
+        call_str += ' -f ' + os.path.join(sub_dir, TIMING_OUT_FILE)
+        call_str += ' -s ' + os.path.join(sub_dir, SOLUTION_OUT_FILE)
+
+        call(call_str, shell=True)
+
+    def _build_param_sets(self):
+        """Returns a list of dictionaries, where the dictionaries are a set of parameters for a single run."""
+        param_keys = self.config.keys()
+        param_values = self.config.values()
+
+        param_combos = list(itertools.product(*param_values))
+
+        param_sets = []
+        for combo in param_combos:
+            param_sets.append({param: value for param, value in zip(param_keys, combo)})
+
+        return param_sets
+
+    @staticmethod
+    def _create_params_str(run_config):
+
```

Comment:
You add whitespace under many functions signatures? If there is a reason for it sure. Personally I am not a fan and I don't see a use for it?

Reply:
I'm playing around with it. Basically it's for me to distinct methods which I do not intend for a user to call from those which a user of the class will call. Not sure though I like the syntax as well. However I think it helps the coder and user of the class.

Reply:
Isn't that the point of the leading underscore? To form some kind of "encapsulation", and let a user of a class know that it is not intended to be called "from the outside"?

Reply:
Exactly. That's why it's in there. The only "public" method so to speak exposed to a user of the class is **run_benchmarks** which has no leading underscore. I should however probably remove them from the property methods. The other stuff is setup things / creating the proper folders / strings etc which is handled internally.

Of course python has no real privacy but I think it's still nice in terms of design and readibility of purpose.

Reply:
Yeah, I was talking about *whitespace* under the signature. You have blank likes under the signature quite often. As in just a line with nothing on it.